cpu/nrf51/periph/pwm.c

```diff
@@ -0,0 +1,224 @@
+/*
+ * Copyright (C) 2018 Freie Universität Berlin
+ *
+ * This file is subject to the terms and conditions of the GNU Lesser General
+ * Public License v2.1. See the file LICENSE in the top level directory for more
+ * details.
+ */
+
+/**
+ * @ingroup     cpu_nrf51
+ * @ingroup     drivers_periph_pwm
+ * @{
+ *
+ * @file
+ * @brief       Low-level PWM driver implementation
+*
+ * @note        This driver initializes a duty cycle of '1'
+ *
+ * @author      Semjon Kerner <semjon.kerner@fu-berlin.de>
+ * @}
+ */
+
+#include <stdint.h>
+#include <string.h>
+#include <math.h>
+
+#include "periph/gpio.h"
+#include "periph/pwm.h"
+#include "vectors_cortexm.h"
+
+#define ENABLE_DEBUG                (1)
+#include "debug.h"
+#define DEBUG_ASSERT_VERBOSE        (1U)
+
+#define NRF51_PWM_PRESCALER_MIN     (0U)
+#define NRF51_PWM_PRESCALER_MAX     (10U)
+#define NRF51_PWM_PPI_CHANNELS      (((1U) << PWM_PPI_A) | ((1U) << PWM_PPI_B))
+#ifndef NRF51_PWM_PERCENT_VAL
+    #define NRF51_PWM_PERCENT_VAL   (1U)
+#endif
+
+inline uint32_t apply_prescaler(uint8_t prescaler)
+{
+    return 16000000 >> prescaler;
+}
+
+/*
+ * compare requested frequency with available frequency
+ * has a margin of PERCENT_VAL
+ */
+bool is_close_to(uint32_t f_requ, uint32_t f_avail)
+{
+    int percent = (f_avail / 100) * NRF51_PWM_PERCENT_VAL;
+    if (f_requ > (f_avail - percent) && f_requ < (f_avail + percent)) {
+        return true;
+    }
+    return false;
+}
+
+uint32_t pwm_init(pwm_t dev, pwm_mode_t mode, uint32_t freq, uint16_t res)
+{
+    assert (dev == 0 && ((mode == PWM_LEFT) || (mode == PWM_RIGHT)));
+
+    /* reset and configure the timer */
+    PWM_TIMER->TASKS_STOP = 1;
+    PWM_TIMER->BITMODE = TIMER_BITMODE_BITMODE_32Bit;
+    PWM_TIMER->MODE = TIMER_MODE_MODE_Timer;
+    PWM_TIMER->TASKS_CLEAR = 1;
+
+    /* calculate and set prescaler */
```

Comment:
looking at this code and thinking about the problem once more, I think we can make this quite more efficient:
```c
// on top of this file define
static const uint32_t divtable[11] = {
    (CLOCK_CPUCLOCK >> 0),
    (CLOCK_CPUCLOCK >> 1),
    ...
    (CLOCK_CPUCLOCK >> 10),
};

// here:
uint32_t lower = (freq - (freq / 100 * NRF51_PWM_PERCENT_VAL));
uint32_t upper = (freq + (freq / 100 * NRF51_PWM_PERCENT_VAL));
for (uint32_t ps = 0; ps <= (PS_MAX + 1); ps++) {
    if (ps == (PW_MAX + 1)) {
        DEBUG("[pwm] init error: resolution or frequency not supported\n");
        return 0;
    }
    if ((divtable[ps] < upper) && (divtable[ps] > lower)) {
        PWM_TIMER->PRESCALER = ps;
        freq = divtable[ps];
        break;
    }
}
...
return freq;
```

Reply:
Indeed this should be the faster solution, but it seems to me as if you confuse pwm frequency and clock frequency at some point. I'll have a look how to reuse your suggestion, thx.